client/lib/viewport/test/react-helpers.js

```diff
@@ -0,0 +1,209 @@
+/**
+ * @format
+ * @jest-environment jsdom
+ */
+
+/**
+ * Internal dependencies
+ */
+import React from 'react';
+import ReactDOM from 'react-dom';
+import { act } from 'react-dom/test-utils';
+
+let helpers;
+
+const listeners = {};
+
+const matchesMock = jest.fn( () => true );
+const addListenerMock = jest.fn( ( query, listener ) => {
+	if ( listeners[ query ] ) {
+		listeners[ query ].push( listener );
+	} else {
+		listeners[ query ] = [ listener ];
+	}
+} );
+const removeListenerMock = jest.fn( ( query, listener ) => {
+	if ( listeners[ query ] ) {
+		listeners[ query ] = listeners[ query ].filter( item => item !== listener );
+	}
+} );
+
+function callQueryListeners( query, value ) {
+	for ( const listener of listeners[ query ] ) {
+		listener( { matches: value } );
+	}
+}
+
+const matchMediaMock = jest.fn( query => {
+	const mediaListObjectMock = {
+		addListener: listener => addListenerMock( query, listener ),
+		removeListener: listener => removeListenerMock( query, listener ),
+	};
+	// Add matches read-only property.
+	Object.defineProperty( mediaListObjectMock, 'matches', {
+		get: () => matchesMock( query ),
+	} );
+	return mediaListObjectMock;
+} );
+
+// Disable console warnings for this file.
+// eslint-disable-next-line no-console
+console.warn = jest.fn();
```

Comment:
that's a global change though isn't it?

Reply:
Doesn't every file run separately in Jest?

Reply:
just saying that this relies on jest isolating globals on test runs across each test file. is that a safe assumption?

Reply:
It does seem potentially dangerous, I agree. I'll find a better way of doing this.

Reply:
Done, using spies now.

Reply:
thanks!

Reply:
> just saying that this relies on jest isolating globals on test runs across each test file. is that a safe assumption?

I agree with doing the mocks as if the globals were not isolated, but it's safe to assume that they are. The global is sandboxed, provided by the [`jest-environment`](https://jestjs.io/docs/en/configuration#testenvironment-string) as configured globally or per-file, and recreated for each file.

Each Jest file can specify a pragma at the beginning:
```js
/**
 * @jest-environment jsdom
 */
```
A rather unavoidable conclusion of this is that the global is different for each file, either Node-ish or browser-ish.